From 4ec9c201068717f72a0a2f5cbb56e14a59984378 Mon Sep 17 00:00:00 2001
From: Neta Zur <neta@marvell.com>
Date: Sun, 30 Aug 2015 08:23:13 +0300
Subject: [PATCH 0376/1240] fix: psci: use generic code for cpu reset address

CPU reset vector address - must be aligned to 0x10000.
use __secure_start - _start as PSCI first address
CPU out of reset first instruction - jump to _armada8k_cpu_entry

Change-Id: I711cfd6b5ab7442e2bdc0c03888880aa9ec3f535
Signed-off-by: Neta Zur <neta@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/23292
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 arch/arm/cpu/armv8/armada8k/psci.S | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/arch/arm/cpu/armv8/armada8k/psci.S b/arch/arm/cpu/armv8/armada8k/psci.S
index 74682ea..202fdbf 100644
--- a/arch/arm/cpu/armv8/armada8k/psci.S
+++ b/arch/arm/cpu/armv8/armada8k/psci.S
@@ -104,12 +104,19 @@ psci_0_2_cpu_on_64:
 
 	/* set the cpu start address */
 	add	x3, x0, #0x640
-	mov	x2, #2 /* WA reset adderrs - CPU reset vector address - must be alined to 0x10000 -
-			    use PSCI first address at 0x20000
-			    first instruction - jump to _armada8k_cpu_entry */
+
+	/* CPU reset vector address - must be aligned to 0x10000
+	use __secure_start - _start as PSCI first address
+	first instruction - jump to _armada8k_cpu_entry */
+	adr     x2, _secure_base
+	ldr     x2, [x2]
+	adr     x4, _start_base
+	ldr     x4, [x4]
+	sub	x2, x2, x4
+	lsr	x2, x2, #16	/* align to 0x10000 */
 	str 	w2, [x3]
 
-	/* save the Linux out of reset address */
+	/* save the Linux out of reset address to unused register */
 	adr	x1, _target_pc
 	ldr	x1, [x1]
 	add	x3, x0, #0x644 /* WA reset adderrs - use reserved register */
@@ -195,5 +202,11 @@ _armada8k_cpu_entry:
 _target_pc:
 	.quad 0x0
 
+_secure_base:
+	.quad __secure_start
+
+_start_base:
+	.quad	_start
+
 text_end:
 	.popsection
-- 
1.9.1

